File: app/src/main/java/me/rei_m/hyakuninisshu/domain/karuta/repository/impl/KarutaExamRepositoryImpl.java
package me.rei_m.hyakuninisshu.domain.karuta.repository.impl;

import android.support.annotation.NonNull;

import com.github.gfx.android.orma.Inserter;
import com.github.gfx.android.orma.SingleAssociation;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;
import me.rei_m.hyakuninisshu.domain.karuta.model.ExamIdentifier;
import me.rei_m.hyakuninisshu.domain.karuta.model.KarutaExam;
import me.rei_m.hyakuninisshu.domain.karuta.model.KarutaExamFactory;
import me.rei_m.hyakuninisshu.domain.karuta.model.KarutaIdentifier;
import me.rei_m.hyakuninisshu.domain.karuta.model.KarutaQuizResult;
import me.rei_m.hyakuninisshu.domain.karuta.repository.KarutaExamRepository;
import me.rei_m.hyakuninisshu.infrastructure.database.ExamWrongKarutaSchema;
import me.rei_m.hyakuninisshu.infrastructure.database.KarutaExamSchema;
import me.rei_m.hyakuninisshu.infrastructure.database.OrmaDatabase;

public class KarutaExamRepositoryImpl implements KarutaExamRepository {

    private final OrmaDatabase orma;

    public KarutaExamRepositoryImpl(@NonNull OrmaDatabase orma) {
        this.orma = orma;
    }

    @Override
    public Maybe<ExamIdentifier> store(@NonNull List<KarutaQuizResult> karutaQuizResultList,
                                       @NonNull Date tookExamDate) {

        KarutaExamSchema karutaExamSchema = new KarutaExamSchema();

        orma.transactionSync(() -> {

            List<KarutaIdentifier> wrongKarutaIdList = new ArrayList<>();

            long totalAnswerTimeMillSec = 0;

            for (KarutaQuizResult karutaQuizResult : karutaQuizResultList) {
                totalAnswerTimeMillSec += karutaQuizResult.answerTime;
                if (!karutaQuizResult.isCollect) {
                    wrongKarutaIdList.add(karutaQuizResult.collectKarutaId);
                }
            }

            final float averageAnswerTime = totalAnswerTimeMillSec / (float) karutaQuizResultList.size() / (float) TimeUnit.SECONDS.toMillis(1);

            karutaExamSchema.tookExamDate = tookExamDate;
            karutaExamSchema.totalQuizCount = karutaQuizResultList.size();
            karutaExamSchema.averageAnswerTime = averageAnswerTime;
            karutaExamSchema.id = KarutaExamSchema.relation(orma).inserter().execute(karutaExamSchema);

            Inserter<ExamWrongKarutaSchema> examWrongKarutaSchemaInserter = ExamWrongKarutaSchema.relation(orma).inserter();
            Observable.fromIterable(wrongKarutaIdList).subscribe(identifier -> {
                ExamWrongKarutaSchema examWrongKarutaSchema = new ExamWrongKarutaSchema();
                examWrongKarutaSchema.examSchema = SingleAssociation.just(karutaExamSchema);
                examWrongKarutaSchema.karutaId = identifier.getValue();
                examWrongKarutaSchemaInserter.execute(examWrongKarutaSchema);
            });
        });

        return Maybe.just(new ExamIdentifier(karutaExamSchema.id));
    }

    @Override
    public Single<KarutaExam> resolve(@NonNull ExamIdentifier identifier) {
        return KarutaExamSchema.relation(orma)
                .idEq(identifier.getValue())
                .selector()
                .executeAsObservable()
                .firstOrError()
                .map(examSchema -> {
                    List<ExamWrongKarutaSchema> examWrongKarutaSchemaList = new ArrayList<>();
                    examSchema.getWrongKarutas(orma)
                            .selector()
                            .executeAsObservable()
                            .subscribe(examWrongKarutaSchemaList::add);
                    return KarutaExamFactory.create(examSchema, examWrongKarutaSchemaList);
                });
    }
}

File: infrastructure/src/main/java/me/rei_m/hyakuninisshu/infrastructure/database/KarutaExamRepositoryImpl.kt
/*
 * Copyright (c) 2018. Rei Matsushita
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

/* ktlint-disable package-name */
package me.rei_m.hyakuninisshu.infrastructure.database

import com.github.gfx.android.orma.SingleAssociation
import me.rei_m.hyakuninisshu.domain.model.karuta.KarutaIdentifier
import me.rei_m.hyakuninisshu.domain.model.karuta.KarutaIds
import me.rei_m.hyakuninisshu.domain.model.quiz.*
import java.util.*

class KarutaExamRepositoryImpl(private val orma: OrmaDatabase) : KarutaExamRepository {

    override fun storeResult(
        karutaExamResult: KarutaExamResult,
        tookExamDate: Date
    ): KarutaExamIdentifier {

        var karutaExamSchemaId: Long = 0

        orma.transactionSync {
            val karutaExamSchema = KarutaExamSchema(
                tookExamDate = tookExamDate,
                totalQuizCount = karutaExamResult.quizCount,
                averageAnswerTime = karutaExamResult.averageAnswerSec
            )
            karutaExamSchemaId = KarutaExamSchema.relation(orma).inserter().execute(karutaExamSchema)
            karutaExamSchema.id = karutaExamSchemaId

            val examWrongKarutaSchemaInserter = ExamWrongKarutaSchema.relation(orma).inserter()
            karutaExamResult.wrongKarutaIds.values.forEach {
                val examWrongKarutaSchema =
                    ExamWrongKarutaSchema(SingleAssociation.just(karutaExamSchema), it.value.toLong())
                examWrongKarutaSchemaInserter.execute(examWrongKarutaSchema)
            }
        }

        return KarutaExamIdentifier(karutaExamSchemaId)
    }

    override fun adjustHistory(historySize: Int) {
        val karutaExamSchemaList = KarutaExamSchema.relation(orma)
            .selector()
            .orderByIdAsc()
            .toList()

        val examCount = karutaExamSchemaList.size

        if (historySize < examCount) {
            orma.transactionSync {
                val deleter = KarutaExamSchema.relation(orma).deleter()
                karutaExamSchemaList.take(examCount - historySize).forEach {
                    deleter.idEq(it.id).execute()
                }
            }
        }
    }

    override fun findBy(karutaExamId: KarutaExamIdentifier): KarutaExam? {
        val examSchema = KarutaExamSchema.relation(orma)
            .idEq(karutaExamId.value)
            .selector()
            .firstOrNull() ?: return null

        val examWrongKarutaSchemaList = examSchema
            .getWrongKarutas(orma)
            .selector()
            .toList()

        return examSchema.toModel(examWrongKarutaSchemaList)
    }

    override fun list(): KarutaExams {
        val examSchemaList = KarutaExamSchema.relation(orma)
            .selector()
            .orderByIdDesc()
            .toList()

        val examList = examSchemaList.map { examSchema ->
            val examWrongKarutaSchemaList = examSchema
                .getWrongKarutas(orma)
                .selector()
                .toList()
            return@map examSchema.toModel(examWrongKarutaSchemaList)
        }

        return KarutaExams(examList)
    }
}

private fun KarutaExamSchema.toModel(examWrongKarutaSchemaList: List<ExamWrongKarutaSchema>): KarutaExam {

    val identifier = KarutaExamIdentifier(id)

    val wrongKarutaIdentifierList = examWrongKarutaSchemaList.map {
        KarutaIdentifier(it.karutaId.toInt())
    }

    val wrongKarutaIds = KarutaIds(wrongKarutaIdentifierList)

    val resultSummary = KarutaQuizzesResultSummary(
        totalQuizCount,
        totalQuizCount - wrongKarutaIds.size,
        averageAnswerTime
    )

    val result = KarutaExamResult(resultSummary, wrongKarutaIds)

    return KarutaExam(identifier, tookExamDate, result)
}
